rdkPlugins/Storage/source/StorageHelper.cpp

```diff
@@ -407,6 +407,33 @@ void StorageHelper::cleanMountLostAndFound(const std::string& mountPoint,
     AI_LOG_FN_EXIT();
 }
 
+// -----------------------------------------------------------------------------
+/**
+ *  @brief Get the loop device for a given backing file
+ *
+ *  @param[in] backingFile   Path of backing file.
+ *
+ *  @return on success a positive file desccriptor corresponding to a free
+ *  loop device, -1 on error.
+ */
+std::string StorageHelper::getLoopDevice(const std::string &backingFile)
+{
+    AI_LOG_FN_ENTRY();
+    
+    const std::string command = "losetup --associated " + backingFile;
```

Comment:
I personally don't like calling out to userspace tools for this, it's prone too breakage if the print output changes and requires that `losetup` is installed on the system.  Also using things like `popen` can expose details we don't want to other tools and results in a full fork and exec.

My suggestion is you just iterate over the files in `/sys/block/loopX/loop/backing_file`, which is what this tool does anyway. ie.
```
int dirFd = open("/sys/block", O_CLOEXEC | O_DIRECTORY);
if (dirFd < 0)
{
   AI_LOG_SYS_ERROR(errno, "failed to open directory @ '/sys/block'");
   return;
}

DIR *sysBlockDir = fdopendir(dirFd);

struct dirent *entry = nullptr;
while ((entry = readdir(sysBlockDir)) != nullptr)
{
   if (strncmp(d->d_name, "loop", 4) != 0)
      continue;

   std::string name(d->d_name);
   name += "/loop/backing_file";

   int fd = openat(dirFd, name.c_str(), O_RDONLY | O_CLOEXEC);
   if (fd < 0)
   {
      AI_LOG_SYS_ERROR(errno, "failed to open file @ '/sys/block/%s'", name.c_str());
   }
   else
   {
      readBackingStoreFile(fd);
      close(fd);
   }
}

closedir(sysBlockDir);
```

Reply:
Yes. You're right. Better not to rely on userspace tools such as losetup. Changed the function to:

`std::string StorageHelper::getLoopDevice(const std::string &backingFile)
{
    AI_LOG_FN_ENTRY();
    
    std::string loopDevice;
    int dirFd = open("/sys/block", O_CLOEXEC | O_DIRECTORY);
    if (dirFd < 0)
    {
        AI_LOG_SYS_ERROR(errno, "failed to open directory @ '/sys/block'");
        return loopDevice;
    }

    DIR *sysBlockDir = fdopendir(dirFd);

    // Loop through all block devices starting with 'loop'
    struct dirent *entry = nullptr;
    while ((entry = readdir(sysBlockDir)) != nullptr)
    {
        if (strncmp(entry->d_name, "loop", 4) != 0)
            continue;

        std::string name(entry->d_name);
        name += "/loop/backing_file";

        int fd = openat(dirFd, name.c_str(), O_RDONLY | O_CLOEXEC);
        if (fd < 0)
        {
            AI_LOG_SYS_ERROR(errno, "failed to open file @ '/sys/block/%s'", name.c_str());
        }
        else
        {
            char buffer[512] = {0};
            int len = read(fd, buffer, sizeof(buffer));
            if (len < 0)
            {
                AI_LOG_SYS_ERROR(errno, "failed to read file @ '/sys/block/%s'", name.c_str());
            }
            else
            {
                if (strncmp(backingFile.c_str(), buffer, backingFile.length()) == 0)
                {
                    loopDevice = std::string("/dev/") + entry->d_name;
                    break;
                }
            }
            close(fd);
        }
    }

    closedir(sysBlockDir);

    AI_LOG_FN_EXIT();
    return loopDevice;
}
`